README.md
# Alien::FFI ![static](https://github.com/PerlFFI/Alien-FFI/workflows/static/badge.svg) ![linux](https://github.com/PerlFFI/Alien-FFI/workflows/linux/badge.svg) ![windows](https://github.com/PerlFFI/Alien-FFI/workflows/windows/badge.svg) ![macos](https://github.com/PerlFFI/Alien-FFI/workflows/macos/badge.svg) ![msys2-mingw](https://github.com/PerlFFI/Alien-FFI/workflows/msys2-mingw/badge.svg)

Build and make available libffi

# SYNOPSIS

In your Makefile.PL:

```perl
use ExtUtils::MakeMaker;
use Alien::Base::Wrapper ();

WriteMakefile(
  Alien::Base::Wrapper->new('Alien::FFI')->mm_args2(
    # MakeMaker args
    NAME => 'My::XS',
    ...
  ),
);
```

In your Build.PL:

```perl
use Module::Build;
use Alien::Base::Wrapper qw( Alien::FFI !export );

my $builder = Module::Build->new(
  ...
  configure_requires => {
    'Alien::FFI' => '0',
    ...
  },
  Alien::Base::Wrapper->mb_args,
  ...
);

$build->create_build_script;
```

# DESCRIPTION

This distribution installs libffi so that it can be used by other Perl distributions.  If already
installed for your operating system, and it can be found, this distribution will use the libffi
that comes with your operating system, otherwise it will download it from the Internet, build and
install it for you.

# SEE ALSO

- [FFI::Platypus](https://metacpan.org/pod/FFI::Platypus)

    Write Perl bindings to non-Perl libraries without C or XS

- [FFI::CheckLib](https://metacpan.org/pod/FFI::CheckLib)

    Check that a library is available for FFI

# AUTHOR

Author: Graham Ollis <plicease@cpan.org>

Contributors:

Petr Písař (ppisar)

# COPYRIGHT AND LICENSE

This software is copyright (c) 2014-2022 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.27      2022-10-23 21:40:22 -0600
  - Download from GitHub (gh#30)

0.26      2022-10-22 10:35:50 -0600
  - Block share install libffi 3.4.3 on Linux Arm64 (gh#29)
  - Move to PerlFFI org

0.25      2020-03-09 03:34:45 -0600
  - Support system install with Visual C++ Perl + Vcpkg (gh#9)

0.24      2019-12-16 08:37:04 -0700
  - Mark libffi 3.3 as bad on windows for a share install.

0.23      2019-03-07 11:47:25 -0500
  - Fix test for big-endian systems.  (gh#7 ppisar++)

0.22      2018-08-28 10:37:44 -0400
  - Additional failure diagnostics

0.21      2018-06-01 07:48:31 -0400
  - Production release identical to 0.20_05

0.20_05   2018-05-30 20:51:57 -0400
  - Diagnostic release

0.20_04   2018-05-30 07:49:31 -0400
  - Diagnostic release

0.20_03   2018-05-29 15:25:11 -0400
  - Added additional tests

0.20_02   2018-05-29 12:41:38 -0400
  - Add some simple minded detection for avx and aes flags on i386
    Only works on systems with /proc/cpuinfo

0.20_01   2018-05-29 10:02:17 -0400
  - Disable avx and aes instructions on i386

0.19      2018-05-19 09:40:13 -0400
  - Identical to 0.18_01 release except for the hardcoded RC0
  - Moved git repository to
    https://github.com/Perl5-FFI/Alien-FFI

0.18_01   2018-04-09 07:51:52 -0400
  - hardcode RC0 for this DEV release only
  - switch from ftp to https, which sourceware seems to provide now

0.17      2017-08-08 15:06:21 -0400
  - Require Alien::Build 0.73 for bug fixes

0.16      2017-03-23 08:00:19 -0400
  - Production release identical to 0.15_02

0.15_02   2017-03-16 21:01:05 -0400
  - Fix regression in libffi version downloaded on windows
  - Fix regressions with Visual C++

0.15_01   2017-03-16 15:42:40 -0400
  - Migrate to Alien::Build + alienfile

0.14      2016-03-01 12:49:45 -0500
  - Use Alien::Base::ModuleBuild for configure_requires instead of Alien::Base

0.12      2015-07-15 16:32:45 -0400
  - Require Alien::Base 0.021 for bug fixes

0.11      2015-04-24 12:41:12 -0400
  - require Alien::Base 0.017 for bug fixes

0.10      2015-04-22 07:20:59 -0400
  - use alien_stage_install

0.09      2015-03-18 13:00:16 -0400
  - Use older version 3.1 when building from source with Microsoft Visual C++
    newer versions seem to be broken with that compiler
  - Build with same -arch flags as Perl on Darwin / OS X

0.08      2015-02-24 12:25:47 -0500
  - Turn off Alien::Base blib detection
  - Make libs and cflags class methods should work correctly
    before or after ./Build install

0.07      2015-02-16 18:16:54 -0500
  - Promote to production

0.06_05   2015-02-16 10:07:27 -0500
  - Fixed MSWin32 logic, removed MB requirement of 0.4202

0.06_04   2015-02-15 08:17:38 -0500
  - Bump MB requirement up to 0.4202

0.06_03   2015-02-14 23:25:21 -0500
  - Only attempt to use Alien::MSYS if it is installed

0.06_02   2015-02-14 08:21:58 -0500
  - Mark prereqs as "dynamic" as we need Alien::MSYS on Windows

0.06_01   2015-02-13 10:43:04 -0500
  - Only require Alien::MSYS if we need to do a build from source

0.06      2015-02-03 15:39:48 -0500
  - Ensure that Alien::Base 0.009 is installed in Build.PL

0.05      2015-01-27 05:03:00 -0500
  - Require Alien::Base 0.009 which supports multiple argument system
    This should relax some of the weirdisms in windows

0.04_01   2015-01-25 22:57:40 -0500
  - work around some weirdisms in windows

0.04      2015-01-23 17:15:50 -0500
  - bump version requirement for Alien::Base 0.007
    I am pretty sure this will correct a cpan testers failure that I am seeing
    in FFI::Platypus.

0.03      2015-01-21 07:30:24 -0500
  - Work-around for unthreaded OpenBSD.

0.02      2014-09-11 16:56:51 -0400
  - make Alien::Base a configure requires

0.01      2014-09-11 14:40:53 -0400
  - initial version


inc/My/ModuleBuild.pm
package My::ModuleBuild;

use strict;
use warnings;
use Config;
use Text::ParseWords qw( shellwords );
use Env qw( @PATH );
use Cwd qw( cwd );
use base qw( Alien::Base::ModuleBuild );

my $libffi_version = '3.2.1';
$libffi_version = '3.1' if $^O eq 'MSWin32' && $Config{ccname} eq 'cl';

# some weird older perl CPAN.pm or Module::Build or something.
# installs okay with older versions of Alien::Base
# even though 0.009 is a configure_requires
# we need to nip it in the bud here else
# it breaks in FFI::Platypus
if($] < 5.010 && ! eval { use Alien::Base 0.009 (); 1 })
{
  print STDERR "Alien::Base 0.009 really is required as a configure prereq.\n";
  print STDERR "Please install it first.\n";
  exit;
}

sub new
{
  my($class, %args) = @_;
  
  my @configure = qw( sh configure --disable-builddir --disable-shared --with-pic --prefix=%s );
  
  if($^O eq 'MSWin32')
  {
    if($Config{archname} =~ /^MSWin32-x64/)
    {
      push @configure, '--build=x86_64-pc-mingw64';
    }
    
    if($Config{ccname} eq 'cl')
    {
      push @configure, qw( --enable-static --disable-shared LD=link ), 'CPP=cl -nologo -EP';
      
      my $cwd = cwd();
      $cwd =~ s{/}{\\\\}g;
      
      if($Config{archname} =~ /^MSWin32-x64/)
      {
        push @configure, "CC=$cwd/_alien/libffi-$libffi_version/msvcc.sh -m64";
      }
      else
      {
        push @configure, "CC=$cwd/_alien/libffi-$libffi_version/msvcc.sh";
      }

    }
  }
  
  if($^O eq 'darwin')
  {
    # perl -MConfig= -MText::ParseWords=shellwords -E 'my @ldflags = shellwords $Config{ldflags}; while(@ldflags) { my $arg = shift @ldflags; next unless $arg eq "-arch"; push @arch, $arg, shift @ldflags }; say $_ for @arch'
    my @ldflags = shellwords $Config{ldflags};
    my @arch;
    while(@ldflags)
    {
      my $arg = shift @ldflags;
      next unless $arg eq "-arch";
      push @arch, $arg, shift @ldflags;
    }
    push @configure, "LDFLAGS=@arch";
    push @configure, "CFLAGS=@arch";
  }
  
  $args{alien_name} = 'libffi';
  $args{alien_build_commands} = [
    \@configure,
    'make',
  ];
  
  if($libffi_version eq '3.1')
  {
    # version 3.1 (only used for MSWin32) generated a bad libffi.pc file
    unshift @{ $args{alien_build_commands} }, [ $^X, '-ni.bak', '-e', 's/^toolexeclibdir=.*$/toolexeclibdir=\$\{libdir\}/; print', 'libffi.pc.in' ];
  }
  
  $args{alien_isolate_dynamic} = 1;
  $args{alien_repository} = {
    protocol => 'http',
    host     => 'mirrors.kernel.org',
    location => '/sourceware/libffi/',
    exact_filename => "libffi-$libffi_version.tar.gz",
  };
  $args{alien_stage_install} = 1;
  
  my $self = $class->SUPER::new(%args);

  if($^O eq 'MSWin32' && ($Alien::Base::ModuleBuild::Force || !$self->alien_check_installed_version))
  {
    $self->_add_prereq( 'build_requires' => 'Alien::MSYS' => 0 );
  }
  
  $self;
}

*do_system = sub
{
  my($self, @rest) = @_;
  require Alien::MSYS;
  my $msys_path = Alien::MSYS::msys_path();
  @PATH = grep { $_ ne $msys_path && $_ !~ /GnuWin32/i } @PATH;
  unshift @PATH, $msys_path;
  $self->SUPER::do_system(@rest);
} if $^O eq 'MSWin32' && eval q{ use Alien::MSYS (); 1 };

1;
